### PR TITLE
Add some tests

### PR DESCRIPTION
**Pull request by Nadrieril**
I was accumulating a list of tests to add, so I thought I'd upstream a batch. I've got quite a few parser tests in particular, that test some corner cases.
I believe this found a bug in the Haskell implementation: in `https://example.com/"a%20b"/c`, the `%20` should _not_ be decoded since it's quoted.
There is also a corner case I could not decide on : should `(merge x y) : T` and  `merge x y : T` be different or the same ? My implementation considers they should be different, which makes typechecking quite easier to implement (essentially a fold). The Haskell implementation however considers them to be the same, which is probably less surprising to the user.

### CODE REVIEW

**Comment by Nadrieril** [comment]
Actually it makes more sense to make those comments into appropriate issues.
Done: https://github.com/dhall-lang/dhall-haskell/issues/932 and https://github.com/dhall-lang/dhall-lang/issues/524

**Comment by singpolyma** [comment]
The IPv6 parse tests fail for me, I assume pending #522

**Comment by Nadrieril** [comment]
@singpolyma I only renamed them, so it's indeed https://github.com/dhall-lang/dhall-lang/pull/522

File: tests/alpha-normalization/success/unit/FunctionNestedBindingXYA.dhall [hunk 1]
```diff
@@ -0,0 +1 @@
+\(_: Bool) -> \(_: Bool) -> _@1 && _
```
Comment (Gabriel439) [inline tests/alpha-normalization/success/unit/FunctionNestedBindingXYA.dhall]: It looks like this test is backwards (i.e. you need to switch the `A.dhall` and `B.dhall` files)

File: tests/alpha-normalization/success/unit/FunctionNestedBindingXXFreeA.dhall [hunk 2]
```diff
@@ -0,0 +1 @@
+\(x: Bool) -> \(x: Bool) -> x && x@0 && x@1 && x@2 && x@3
```
Comment (Gabriel439) [inline tests/alpha-normalization/success/unit/FunctionNestedBindingXXFreeA.dhall]: This caught a bug in the Haskell implementation, too 🙂

File: tests/alpha-normalization/success/unit/FunctionNestedBindingXYB.dhall [hunk 3]
```diff
@@ -0,0 +1 @@
+\(x: Bool) -> \(y: Bool) -> x && y
```
Comment (philandstuff) [inline tests/alpha-normalization/success/unit/FunctionNestedBindingXYB.dhall]: This test works, but doesn't match the README which says that the `B` version should be the normalized version.  I feel we should at least be consistent.  README quoted below:

```md
### Running `alpha-normalization` tests

The tests should:
- parse `A` and `B`
- α-normalize them both
- the results should match

Where `A` and `B` are:
- `A`: unnormalized text
- `B`: normalized text
```

File: tests/parser/failure/unit/BoolLitTrueWithIndex.dhall [hunk 4]
```diff
@@ -0,0 +1 @@
+Bool@0
```
Comment (philandstuff) [inline tests/parser/failure/unit/BoolLitTrueWithIndex.dhall]:
```suggestion
True@0
```
Based on the filename, I think this is what you intended?

File: tests/parser/failure/unit/ListLitEmptyPriority.dhall [hunk 5]
```diff
@@ -0,0 +1 @@
+[] : List T U
```
Comment (philandstuff) [inline tests/parser/failure/unit/ListLitEmptyPriority.dhall]: What does "Priority" mean in this file name?  Is it worth a comment explaining what's going on here?  (Do you perhaps mean "precedence"?)

File: tests/parser/success/unit/LambaA.dhall [hunk 6]
```diff
@@ -0,0 +1 @@
+\(x: T) -> x
```
Comment (philandstuff) [inline tests/parser/success/unit/LambaA.dhall]: While "lamda" is a variant spelling, we use "lambda" everywhere else and we should probably be consistent.

File: tests/parser/success/unit/MergeAnnotationPriorityA.dhall [hunk 7]
```diff
@@ -0,0 +1 @@
+merge x y : t a
```
Comment (philandstuff) [inline tests/parser/success/unit/MergeAnnotationPriorityA.dhall]: Again, "priority" maybe should be "precedence"? Not sure what you intended though.

File: tests/parser/success/unit/MergeParenAnnotationA.dhall [hunk 8]
```diff
@@ -0,0 +1 @@
+(merge x y) : t
```
Comment (philandstuff) [inline tests/parser/success/unit/MergeParenAnnotationA.dhall]: (I can't comment on binary files directly.) Above there's a MergeB.dhallb file but no MergeA.dhall file. Did you forget to add it?

File: tests/parser/success/unit/QuotedVariableA.dhall [hunk 9]
```diff
@@ -0,0 +1 @@
+`x`
```
Comment (philandstuff) [inline tests/parser/success/unit/QuotedVariableA.dhall]: is it worth putting in some characters that must be quoted here, to test that you can actually use nonstandard characters in quotes? eg `` `x+y` `` or `` `f♯` ``

File: tests/parser/success/unit/UnionTypeXTYTA.dhall [hunk 10]
```diff
@@ -0,0 +1 @@
+< x: T | y: U >
```
Comment (philandstuff) [inline tests/parser/success/unit/UnionTypeXTYTA.dhall]: It's a very small thing, but shouldn't the filename be UnionTypeXTYUA.dhall?

File: tests/normalization/success/unit/OperatorTextConcatenateRhsEmptyB.dhall [hunk 11]
```diff
@@ -0,0 +1 @@
+"${x}"
```
Comment (Gabriel439) [inline tests/normalization/success/unit/OperatorTextConcatenateRhsEmptyB.dhall]: Note that now #515 is merged this should be `x`

File: tests/parser/success/unit/import/urls/quotedPathFakeUrlEncodeA.dhall [hunk 12]
```diff
@@ -0,0 +1 @@
+https://example.com/"a%20b"/c
```
Comment (philandstuff) [inline tests/parser/success/unit/import/urls/quotedPathFakeUrlEncodeA.dhall]: (urgh, I can't comment on a renamed file)
You moved pathTerminationA.dhall  into the `parser/success/unit/import` directory, but I don't think it's strictly a unit test... it can be broken down into separate cases per delimiter.  (In particular, I haven't implemented unions, so I can't run the test, though I would benefit from the rest of it)

File: tests/parser/failure/unit/ListLitEmptyPriority.dhall (hunk 5 above)
Comment (Nadrieril) [reply tests/parser/failure/unit/ListLitEmptyPriority.dhall]: I did mean precedence, thanks :)

File: tests/parser/success/unit/LambaA.dhall (hunk 6 above)
Comment (Nadrieril) [reply tests/parser/success/unit/LambaA.dhall]: I had not idea it was a valid spelling, that was definitely a typo

File: tests/parser/success/unit/MergeParenAnnotationA.dhall (hunk 8 above)
Comment (Nadrieril) [reply tests/parser/success/unit/MergeParenAnnotationA.dhall]: It's somewhere below for some reason

Comment (philandstuff) [reply tests/parser/success/unit/MergeParenAnnotationA.dhall]: Oh yeah, it was just a weird file ordering thing because of intervening `MergeAnnotation...` files

**Comment by Nadrieril** [comment]
Thanks for the precise review !
